Added option to replace dots in keys with underscores

MarcusNoble opened the pull request:
*Issue #, if available:* N/A

*Description of changes:*
Add an option to replace dots (`.`) within keys with underscores (`_`) to help avoid type clashes in Elasticsearch and similar.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

PettitWesley commented on kinesis/kinesis.go:
```diff
@@ -185,6 +189,32 @@ func (outputPlugin *OutputPlugin) Flush() error {
+		case map[interface{}]interface{}:
+			v = replaceDots(vt)
+			// Ensure duplicated keys are merged rather than replaced
+			if _, ok := newObj[k]; ok {
```
checking for duplicated keys needs to happen in the first call to `replaceDots`, not just in the recursive case

PettitWesley replied:
Interestingly, the C code for the ES version of this option is very simple and does not check for duplicates:

https://github.com/fluent/fluent-bit/blob/master/plugins/out_es/es.c#L90

PettitWesley replied:
Checking for duplicated keys is safer; I like having that check. However, I think it should be able to be done in place without allocating a new map.

I believe this should be safe in Go: https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-map-within-a-range-loop

PettitWesley commented on kinesis/kinesis.go:
```diff
@@ -185,6 +189,32 @@ func (outputPlugin *OutputPlugin) Flush() error {
 }
 
+func replaceDots(obj map[interface{}]interface{}) map[interface{}]interface{} {
+	newObj := make(map[interface{}]interface{})
```
I'd like this code to work without requiring an additional map allocation... that's going to add a lot overhead.

MarcusNoble commented:
Thanks for your review @PettitWesley. 

I've removed the re-assignment so now it just does a mutation in place. 

I've also removed the deduplication code as that needs some more thought. I'm going to open an issue to discuss it and provide the scenario that I'm hitting a problem.

PettitWesley commented:
@MarcusNoble I lean towards saying this should be a feature in Fluent Bit core; see here: https://github.com/fluent/fluent-bit/issues/1305#issuecomment-585993039

nsvijay04b1 commented:
i am one of those who are waiting for this merge so i can use it :-)

nsvijay04b1 commented:
hi .. Any progress please? awaiting fix as i have the same problem.

nsvijay04b1 commented:
Release v2.7.0  also missed this feature.   Hope this will be part of Release v2.8.0

hossain-rayhan commented:
Hi @nsvijay04b1 , as @PettitWesley  mentioned, this might a good option to put into the `fluent-bit` core. So, I am not sure if we are planning to implement it separately for each of our plugins. 

Check this for update: https://github.com/fluent/fluent-bit/issues/1305#issuecomment-585993039

PettitWesley commented:
We're now thinking of working on this and including it in an upcoming release

PettitWesley commented:
@MarcusNoble @nsvijay04b1 We have take this back up in #79. Marcus's original commits are preserved.

Apologies for taking so long to deliberate on this.
